lib/srv/usermgmt.go

```diff
@@ -227,23 +229,67 @@ func (u *HostUserManagement) CreateUser(name string, ui *services.HostUsersInfo)
 		return nil, trace.BadParameter("Mode is a required argument to CreateUser")
 	}
 
+	groups := make([]string, 0, len(ui.Groups))
+	for _, group := range ui.Groups {
+		if group == name {
+			// this causes an error as useradd expects the group with the same name as the user to be available
+			log.Debugf("Skipping group creation with name the same as login user (%q, %q).", name, group)
+			continue
+		}
+		groups = append(groups, group)
+	}
+	if ui.Mode == types.CreateHostUserMode_HOST_USER_MODE_INSECURE_DROP {
+		groups = append(groups, types.TeleportServiceGroup)
+	}
+	var errs []error
+	for _, group := range groups {
+		if err := u.createGroupIfNotExist(group); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+	if err := trace.NewAggregate(errs...); err != nil {
+		return nil, trace.WrapWithMessage(err, "error while creating groups")
+	}
+
 	tempUser, err := u.backend.Lookup(name)
 	if err != nil && !errors.Is(err, user.UnknownUserError(name)) {
 		return nil, trace.Wrap(err)
 	}
 
 	if tempUser != nil {
-		gids, err := u.backend.UserGIDs(tempUser)
-		if err != nil {
-			return nil, trace.Wrap(err)
+		// Collect actions that need to be done together under a lock on the user.
+		actionsUnderLock := []func() error{
+			func() error {
+				// If the user exists, set user groups again as they might have changed.
+				return trace.Wrap(u.backend.SetUserGroups(name, groups))
+			},
+		}
+		doWithUserLock := func() error {
+			return trace.Wrap(u.doWithUserLock(func(_ types.SemaphoreLease) error {
+				for _, action := range actionsUnderLock {
+					if err := action(); err != nil {
+						return trace.Wrap(err)
+					}
+				}
+				return nil
+			}))
 		}
+
 		systemGroup, err := u.backend.LookupGroup(types.TeleportServiceGroup)
 		if err != nil {
 			if isUnknownGroupError(err, types.TeleportServiceGroup) {
+				if err := doWithUserLock(); err != nil {
+					return nil, trace.Wrap(err)
+				}
```

Comment:
Relatedly, we never seem to call `UpsertHostUserInterationTime` for 2nd-time `KEEP` users, I think? Is that intended? I guess that just controls the deletion timer so maybe it's weird that we create the record at all for `KEEP` users.

Reply:
We call `doWithUserLock()` because the groups need to be updated even if the user exists. But I agree that it's weird to return an error for that if we just ignore it. I've renamed `CreateUser` to `UpsertUser`, which will hopefully make it clearer that a user already existing isn't an error case.

For interaction time, I think you're right that keep users don't need it. I've updated the code to only set it for non-keep users.

Reply:
Thanks! It reads much more clearly now.